File: templates/python/code-carl/ch06_tree/t9_lc0111_min_depth.py
"""LeetCode 111. Minimum Depth of Binary Tree
https://leetcode.cn/problems/minimum-depth-of-binary-tree/
"""

from typing import Optional

from tree_node import TreeNode


class Solution:
    def minDepth(self, root: Optional[TreeNode]) -> int:
        pass


if __name__ == "__main__":
    root1 = TreeNode(3, TreeNode(9), TreeNode(20, TreeNode(15), TreeNode(7)))
    solution = Solution()
    assert solution.minDepth(root1) == 2

    root2 = TreeNode(
        2, None, TreeNode(3, None, TreeNode(4, None, TreeNode(5, None, TreeNode(6))))
    )
    solution = Solution()
    assert solution.minDepth(root2) == 5

    print("All tests passed.")

File: solutions/code-carl/ch06_tree/T9_LC0111_MinDepth.java
/**
 * Practice template for:
 * - LeetCode 111. Minimum Depth of Binary Tree
 *   https://leetcode.cn/problems/minimum-depth-of-binary-tree/
 */
public class T9_LC0111_MinDepth {
    public int minDepth(TreeNode root) {
        if (root == null) {
            return 0;
        }
        int left = minDepth(root.left);
        int right = minDepth(root.right);
        if (left == 0) {
            return right + 1;
        } else if (right == 0) {
            return left + 1;
        } else {
            return Math.min(left, right) + 1;
        }
    }
}
